pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -224,3 +234,425 @@ func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32,
 
 	return replicas, limited, nil
 }
+
+func applyCounterOrListPolicy(c *autoscalingv1.CounterPolicy, l *autoscalingv1.ListPolicy,
+	f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
+	nodeCounts map[string]gameservers.NodeCount) (int32, bool, error) {
+
+	if !runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
+		return 0, false, errors.Errorf("cannot apply CounterPolicy unless feature flag %s is enabled", runtime.FeatureCountsAndLists)
+	}
+
+	var isCounter bool    // True if a CounterPolicy False if a ListPolicy
+	var key string        // The specified Counter or List
+	var count int64       // The Count or number of Values in the template Game Server
+	var capacity int64    // The Capacity in the template Game Server
+	var aggCount int64    // The Aggregate Count of the specified Counter or List of all GameServers across the GameServerSet in the Fleet
+	var aggCapacity int64 // The Aggregate Capacity of the specified Counter or List of all GameServers across the GameServerSet in the Fleet
+	var minCapacity int64 // The Minimum Aggregate Capacity
+	var maxCapacity int64 // The Maximum Aggregate Capacity
+
+	var bufferSize intstr.IntOrString
+	// TODO: Our aggregate counts in include Allocated, Ready, and Reserved replicas, however
+	// I'm not 100% sure if f.Status.Replicas includes all three states or just Allocated and Ready?
+	// If it's the latter, then we would need to add in f.Status.ReservedReplicas.
+	// There may be some additional nuance here since "Reserved instances won't be deleted on scale down, but won't cause an autoscaler to scale up"
+	replicas := f.Status.Replicas
```

Comment:
Hmm, should we update AggregateCounts and AggregateCapacity to just include Allocated + Ready? And maybe have a second field for reserved? Or three separate fields for AllocatedAggregateCount, ReadyAggregateCount, ReservedAggregateCount?

Reply:
I'm leaning to the later to keep consistent with the replica fields, and be more flexible.

Reply:
To counter my own argument - when I did player tracking, I only looked at Ready/Reserved/Allocated

https://github.com/googleforgames/agones/blob/cef40425076e5e9a766952eb2f499d9b447b703a/pkg/gameserversets/controller.go#L630-L648

But I also never got to doing autoscaling based on player tracking (which for us is what the aggregates are most used for), so it's possible that would have changed over time, since you may well want to count the capacity that is incoming form a yet-to-be-Ready-GameServer, so you don't overshoot on resources and then have to scale it back (which is why I would recommend counting across all GameServers in the gsSet + Fleet) 🤔  I do wonder why I did that.

Reply:
I'll write this another way as well -- I assume that when we get into autoscaling, there may be extra status values above and beyond what we had in https://github.com/googleforgames/agones/issues/2716#:~:text=1000%20item%20capacity-,Status,-status%3A%0A%20%23 that are useful to us. I figure we can also add them then, so we don't have to get this 100% right in this round.

Reply:
It seems like if we are trying to abide by the "Reserved instances won't be deleted on scale down, but won't cause an autoscaler to scale up" then the aggregate values should not include reserved gameservers, as it stands this autoscaling logic includes reserved counts. If we're not abiding by that logic for Counts and Lists then we should probably make that clear? (And fwiw, we're technically tie-breaking on packed node sorting for any gameserver that is not in an "IsBeforeReady" state in sortGameServersByLeastFullNodes.)

Reply:
Per conversation:

We change from current calculation of AggregateCount and Capacity for Ready, Reserved, or Allocated only to AggregateCounts and AggregateCapacity is across all statuses (like Replicas). This will not include game servers that are being shut down (as is the current state for status.Replicas).

We add an AllocatedAggregateCounts and AllocatedAggregateCapacity that we use to calculate the appropriate buffer at the buffer should be determined based on allocated available capacity.